designs/DeveloperTestingPackageExperience.md

```diff
@@ -0,0 +1,125 @@
+# Developer Testing Package Experience
+
+* Status: In Review
+* Author(s): [Samuel Cragg](https://github.com/samcragg)
+* Issue: [6579](https://github.com/NuGet/Home/issues/6579) Enhance the
+  experience for the testing of packages in development
+
+## Problem Background
+
+As a NuGet package author, it would be nice to be able to test the package
+locally before publishing to a NuGet feed.
+
+As packages can include custom MsBuild targets, it's important to verify that
+the generated package works as expected when restored into a project. Being able
+to verify that locally (including on a CI server) will enable developers to spot
+changes in behaviour sooner, enhancing the inner-loop performance for .NET
+developers.
+
+## Who are the customers
+
+NuGet package authors (i.e. developers who create NuGet packages).
+
+## Goals
+
+The goal of this design is to enable a generated `nupkg` file to be used by a
+project using the existing `PackageReference` functionality. The version of the
+package may be an existing published version or that of a previously restored
+local package (i.e. there is no requirement that the version is unique, as there
+is with the current package publishing requirements), however, it is assumed
+that the package ID will be unique to avoid conflict with other dependencies.
+
+## Non-Goals
+
+Custom feeds will not be supported by the solution - the solution is only
+interested in folders containing `nupkg` files. The rationale for this is to
+keep the design simple and to emphasise that it is not a replacement for other
+package sources. Also, since the target audience is creating the packages, the
+file would have been produced in order to publish it, so this limitation is not
+perceived as an impediment.
+
+## Solution Overview
+
+In order to provide a solution, there are a few issues that need to be
+highlighted with the current machinery with regards to installing of a package
+in the context of trying to test a newly created package:
+
+* NuGet caches packages by ID and version
+* The order of multiple sources are searched through is non-deterministic
+* If a version number is omitted, NuGet searches for the lowest applicable
+  version
+
+### `csproj` Additions
+
+The proposed solution is to add an additional MsBuild property for the
+[restore target](https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restore-target)
+as follows:
+
+| Property              | Description  |
+|-----------------------|--------------|
+| RestoreOverrideFolder | Specifies a folder that contains `nupkg` files that take precedence over any other source and are not cached. |
+
+This property can only be specified once and can only specify a single folder
+(i.e. there cannot be multiple override folders to avoid ordering issues). Note
+that file shares _should_ be supported, however.
+
+Also, only a single version of a package (based on its ID) may exist in the
```

Comment:
Is the idea to overwrite *many* packages at once?

Reply:
How do we handle transitive dependencies?
Would those be expected to be in that local folder? 

I think package versions *have* to matter for transitive dependencies, because this behavior breaks when you have 3rd party dependencies to the package being developed.

Reply:
I was hoping this would be handled for free by limiting it to a single version of that package in the hypothetical override folder. When looking for dependencies, the same logic would apply, look in the override folder first and then fallback to others. I don't think versioning would matter in this case - if you've specified a version and it doesn't exist in the override folder then the normal logic applies. If you specify a version that does exist in the override folder, then that is used without checking any other sources. If you're not using an exact version but the version that exists in the override folder exists fits the bill, it will be used.

Bear in mind the override folder is really only meant to be used for testing a package that you've just built locally. I don't think other uses should be supported as it will add to complexity. But obviously, my opinion is biased 🙂